src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py

```diff
@@ -0,0 +1,431 @@
+    ):
+ 
+        down_block_res_samples, mid_block_res_sample = None, None
+
+        noise_pred = self.unet(
+            latent_model_input,
+            timestep=t,
+            encoder_hidden_states=text_embeddings,
+            down_block_additional_residuals=down_block_res_samples,
+            mid_block_additional_residual=mid_block_res_sample,
+        )["sample"]
+
+        return noise_pred
+
+    def _init_latent(self, latent, height, width, generator, batch_size):
+        if latent is None:
+            latent = torch.randn(
+                (1, self.unet.in_channels, height // 8, width // 8),
+                generator=generator,
+                device=generator.device,
+            )
+        latent = latent.expand(
+            batch_size,
+            self.unet.in_channels,
+            height // 8,
+            width // 8,
+        )
+        return latent.to(self.device)
+
+    def _forward_prompt_embeddings(self, input_image, src_subject, prompt):
+        # 1. extract BLIP query features and proj to text space -> (bs, 32, 768)
+        query_embeds = self.forward_ctx_embeddings(input_image, src_subject)
+
+        # 2. embeddings for prompt, with query_embeds as context
+        tokenized_prompt = self._tokenize_text(prompt).to(self.device)
+        text_embeddings = self.text_encoder(
+            input_ids=tokenized_prompt.input_ids,
+            ctx_embeddings=query_embeds,
+            ctx_begin_pos=[self._CTX_BEGIN_POS],
+        )[0]
+
+        return text_embeddings
+
+    @torch.no_grad()
+    def get_image_latents(self, image, sample=True, rng_generator=None):
+        assert isinstance(image, torch.Tensor)
+
+        encoding_dist = self.vae.encode(image).latent_dist
+        if sample:
+            encoding = encoding_dist.sample(generator=rng_generator)
+        else:
+            encoding = encoding_dist.mode()
+        latents = encoding * 0.18215
+        return latents
+
+    def preprocess_caption(self, caption):
+        caption = re.sub(
+            r"([.!\"()*#:;~])",
+            " ",
+            caption.lower(),
+        )
+        caption = re.sub(
+            r"\s{2,}",
+            " ",
+            caption,
+        )
+        caption = caption.rstrip("\n")
+        caption = caption.strip(" ")
+
+        # truncate caption
+        caption_words = caption.split(" ")
+        if len(caption_words) > 50:
+            caption = " ".join(caption_words[:50])
+
+        return caption
+    
+    def preprocess_image(self, image):
+                
+        mean = (0.48145466, 0.4578275, 0.40821073)
+        std = (0.26862954, 0.26130258, 0.27577711)
+        transform = transforms.Compose(
+            [
+                transforms.Resize(224, interpolation=InterpolationMode.BICUBIC),
+                transforms.CenterCrop(224),
+                transforms.ToTensor(),
+                transforms.Normalize(mean, std)
+            ]
+        )
+
+        return transform(image)
```

Comment:
We should not introduce a dependency on `torchvision` here and instead adapt: 

https://github.com/huggingface/diffusers/blob/078df46bc9a99178a9a744b872899990353769a4/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_instruct_pix2pix.py#L296

Cc: @yiyixuxu for refactoring needed to support this preprocessing.

Reply:
Thanks for explaining. 

Yeah let's do that in the image_processor.py module.